deploy/deploy.go

```diff
@@ -0,0 +1,582 @@
+		var rte *route
+		for _, r := range routes {
+			if r.routeRE.MatchString(path) {
+				rte = r
+				break
+			}
+		}
+		lf, err := newLocalFile(fs, path, rte)
+		if err != nil {
+			return err
+		}
+		retval[path] = lf
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return retval, nil
+}
+
+// walkRemote walks the target bucket and returns a flat list.
+func walkRemote(ctx context.Context, bucket *blob.Bucket) (map[string]*blob.ListObject, error) {
+	retval := map[string]*blob.ListObject{}
+	iter := bucket.List(nil)
+	for {
+		obj, err := iter.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		retval[obj.Key] = obj
+	}
+	return retval, nil
+}
+
+// uploadReason is an enum of reasons why a file must be uploaded.
+type uploadReason string
+
+const (
+	reasonUnknown    uploadReason = "unknown"
+	reasonNotFound   uploadReason = "not found at target"
+	reasonForce      uploadReason = "--force"
+	reasonSize       uploadReason = "size differs"
+	reasonMD5Differs uploadReason = "md5 differs"
+	reasonMD5Missing uploadReason = "remote md5 missing"
+)
+
+// fileToUpload represents a single local file that should be uploaded to
+// the target.
+type fileToUpload struct {
+	Local  *localFile
+	Reason uploadReason
+}
+
+func (u *fileToUpload) String() string {
+	details := []string{humanize.Bytes(uint64(u.Local.UploadSize))}
+	if s := u.Local.CacheControl(); s != "" {
+		details = append(details, fmt.Sprintf("Cache-Control: %q", s))
+	}
+	if s := u.Local.ContentEncoding(); s != "" {
+		details = append(details, fmt.Sprintf("Content-Encoding: %q", s))
+	}
+	if s := u.Local.ContentType(); s != "" {
+		details = append(details, fmt.Sprintf("Content-Type: %q", s))
+	}
+	return fmt.Sprintf("%s (%s): %v", u.Local.Path, strings.Join(details, ", "), u.Reason)
+}
+
+// findDiffs diffs localFiles vs remoteFiles to see what changes should be
+// applied to the remote target. It returns a slice of *fileToUpload and a
+// slice of paths for files to delete.
+func findDiffs(localFiles map[string]*localFile, remoteFiles map[string]*blob.ListObject, force bool) ([]*fileToUpload, []string) {
+	var uploads []*fileToUpload
+	var deletes []string
+
+	// TODO: Do we need to remap file delimiters, e.g. on Windows?
+
+	found := map[string]bool{}
+	for path, lf := range localFiles {
+		upload := false
+		reason := reasonUnknown
+
+		if remoteFile, ok := remoteFiles[path]; ok {
+			// The file exists in remote. Let's see if we need to upload it anyway.
+
+			// TODO: We don't register a diff if the metadata (e.g., Content-Type
+			// header) has changed. This would be difficult/expensive to detect; some
+			// providers return metadata along with their "List" result, but others
+			// (notably AWS S3) do not, so gocloud.dev's blob.Bucket doesn't expose
+			// it in the list result. It would require a separate request per blob
+			// to fetch. At least for now, we work around this by documenting it and
```

Comment:
Why not extract a sample set of 1 file per extension + 1 file per route match and check those files? It should be a small set of files. It would only work if you use `hugo deploy` only, but that is a trade off I'm willing to take.

Reply:
That's a good idea. I added it to the TODO.